### PR TITLE
Fix/minimize rebuild

### PR DESCRIPTION
**Pull request by gammazero**
This PR builds on, and is a replacement for PR #13.

This PR addresses two issues with the pinner:
1. An untended shutdown of IPFS can lead to saving pins in an incomplete state, requiring rebuilding indexes at the next start up.
2. A defect in the pin index rebuild process caused all indexes to be rebuilt, whether needed or not.

These solutions to these problems:
1. Pins are automatically synced after each modification to pins and indexes.  This default behavior can be disabled and re-enabled if needed for high-volume operations.  Additionally, syncing pins is done separately from syncing the dag service, so a long duration dag service sync does prolong the time that pins are in an incomplete "dirty" state.
2. The rebuild process fixes a problem with setting the correct ID when loading pins.  Additionally, large numbers of pin and index data are not loaded into maps during this process.

Additional tests have been added to test the correctness of indexes and the rebuild process.

### CODE REVIEW

**Review by petar — COMMENTED** [review]
It looks reasonable. The highest value here appears to be the more efficient index rebuilding. This pinner implementation is becoming hard to review and verify. Going forward it should teach us to do the right thing: use transactions or write-ahead logging. I am suspecting that every bug fix iteration on this implementation is comparable in time to re-implementing the whole thing using transactions. In any event, this looks better than the previous version!

File: dspinner/pin.go [hunk 1]
```diff
@@ -914,26 +921,127 @@ func decodePin(pid string, data []byte) (*pin, error) {
...
+		} else if pp.Mode == ipfspinner.Direct {
+			indexer = p.cidDIndex
+		} else {
+			log.Error("unrecognized pin mode:", pp.Mode)
+			continue
+		}
+
+		// Check that the indexer indexes this pin
+		indexKey := pp.Cid.KeyString()
+		ok, err := indexer.HasValue(ctx, indexKey, pp.Id)
```
[the remark this answers is not in the document]
Comment (gammazero) [reply dspinner/pin.go:997]: Yes, that may be a possibility given that the datastore can flush in an indeterminate order.  This is now handled.

File: dspinner/pin.go [hunk 2]
```diff
@@ -914,26 +921,127 @@ func decodePin(pid string, data []byte) (*pin, error) {
...
+func (p *pinner) setDirty(ctx context.Context) {
+	wasClean := p.dirty == p.clean
+	p.dirty++
+
+	if !wasClean {
+		return // do not save; was already dirty
+	}
+
+	data := []byte{1}
+	err := p.dstore.Put(dirtyKey, data)
```
[the remark this answers is not in the document]
Comment (gammazero) [reply dspinner/pin.go:912]: It is only read at startup, and the counter might as well reset then. 🤷

File: dspinner/pin.go [hunk 3]
```diff
@@ -810,37 +796,55 @@ func (p *pinner) Update(ctx context.Context, from, to cid.Cid, unpin bool) error
...
 	}
+	p.setClean(ctx)
+	return nil
+}
 
-	p.setDirty(ctx, false)
+// Flush encodes and writes pinner keysets to the datastore
+func (p *pinner) Flush(ctx context.Context) error {
+	// If autosync is enabled, then this is not needed.
+	if p.autoSync {
```
[the remark this answers is not in the document]
Comment (gammazero) [reply dspinner/pin.go]: Good catch. Fixed.

File: dspinner/pin.go [hunk 4]
```diff
@@ -244,9 +265,33 @@ func (p *pinner) addPin(ctx context.Context, c cid.Cid, mode ipfspinner.Mode, na
 		return "", fmt.Errorf("could not encode pin: %v", err)
 	}
 
-	p.setDirty(ctx, true)
+	p.setDirty(ctx)
+
+	// Store the pin.  Pin is stored first so that an incomplete add is
+	// detected by a pin that does not have an index.
+	err = p.dstore.Put(pp.dsKey(), pinData)
```
[the remark this answers is not in the document]
Comment (gammazero) [reply dspinner/pin.go:271]: True, the underlying datastore can order the flushing of data in any way.  Removed the comment.

Really, using transactions would avoid any mystery here.

File: dspinner/pin.go [hunk 5]
```diff
@@ -244,9 +265,33 @@ func (p *pinner) addPin(ctx context.Context, c cid.Cid, mode ipfspinner.Mode, na
...
+	if err != nil {
+		if mode == ipfspinner.Recursive {
```
[the remark this answers is not in the document]
Comment (gammazero) [reply dspinner/pin.go]: I don't.  That was left over from when the pin was written first.

Fixed.

**Comment by gammazero** [comment]
Using transactions for this is the better fix.  However, this means that the pinner must be given a transaction-capable datastore, and the pinner will need to be able to get the transaction interface from the datastore --  which is not difficult to solve. However, if there are pinners currently using datastores that do not support transactions, then this will require a migration to a new datastore, and that may require a larger scoped change.

**Review by petar — APPROVED** [review]
LGTM.

@aschmahmann should we deploy in the clusters before merging to master?

**Comment by aschmahmann** [comment]
Running some live tests here seems reasonable to me if we're pretty confident that things should work and we're close to merge (i.e. we don't want to mess up the cluster nodes if we can avoid it). I would only deploy to one at a time in case there are issues.

As for transactions. I wouldn't worry too much about little using non-transactional datastores, although we can take care of this in a subsequent PR. The main datastores that I know of that are in use here are LevelDB and Badger both of which should support transactions. Maybe there's also a map datastore, but we could either make one that supports transactions or just use the in memory levelDB datastore.

File: dspinner/pin.go [hunk 6]
```diff
@@ -981,15 +981,15 @@ func (p *pinner) rebuildIndexes(ctx context.Context) error {
 			indexer = p.cidRIndex
 			// Delete any direct index
 			err = p.cidDIndex.Delete(ctx, indexKey, pp.Id)
-			log.Infof("deleting stale pin index for cid %v", pp.Cid.String())
+			log.Errorf("deleting stale pin index for cid %v", pp.Cid.String())
```
Comment (gammazero) [inline dspinner/pin.go:984]: This appears to log the error message whether or not a stale pin was actually found.  Perhaps `Delete` needs to return a boolean, or there should be a find for the pin first.

Comment (petar) [reply dspinner/pin.go:984]: Let's not log deletions for now. I'll update the PR.

**Pull request merged**